File: problems/2815-max-pair-sum-in-an-array/solution.py
class Solution:
    def maxSum(self, nums: List[int]) -> int:
        def largest_digit(n):
            max_digit = 0
            while n > 0:
                digit = n % 10
                max_digit = max(max_digit, digit)
                n //= 10
            return max_digit

        max_sum = -1
        for i in range(len(nums)):
            for j in range(i + 1, len(nums)):
                if largest_digit(nums[i]) == largest_digit(nums[j]):
                    max_sum = max(max_sum, nums[i] + nums[j])

        return max_sum

File: problems/2815-max-pair-sum-in-an-array/solution.java
class Solution {
    public int maxSum(int[] nums) {
        int maxSum = -1;
        int[] maxDigitMap = new int[10];
        for (int num : nums) {
            int maxDigit = 0;
            int temp = num;
            while (temp > 0) {
                maxDigit = Math.max(maxDigit, temp % 10);
                temp /= 10;
            }
            if (maxDigitMap[maxDigit] != 0) {
                maxSum = Math.max(maxSum, maxDigitMap[maxDigit] + num);
            }
            maxDigitMap[maxDigit] = Math.max(maxDigitMap[maxDigit], num);
        }
        return maxSum;
    }
}